// react-magic-grid/dist/react-magic-grid.es.js
import React, { forwardRef, useRef, useEffect, useImperativeHandle, createRef } from 'react';
import PropTypes from 'prop-types';
import MagicGrid from 'magic-grid';

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

var MagicGridWrapper = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"]);

  var container = useRef(null);
  var grid = /*#__PURE__*/createRef();
  useEffect(function () {
    var timeout; // magic-grid handles resizing via its own `listen` method
    // unfortunately event listener it creates is not being cleaned up
    // that's why we don't use it and have our own instead
    // see: https://github.com/e-oj/Magic-Grid/issues/24

    var resize = function resize() {
      if (!timeout) timeout = setTimeout(function () {
        grid.current && grid.current.positionItems();
        timeout = null;
      }, 200);
    };

    if (!grid.current) {
      grid.current = new MagicGrid(_objectSpread2({
        container: container.current
      }, props));
      window.addEventListener("resize", resize);
    }

    grid.current.positionItems();
    return function () {
      window.removeEventListener("resize", resize);
    };
  });
  useImperativeHandle(ref, function () {
    return {
      positionItems: function positionItems() {
        grid.current && grid.current.positionItems();
      }
    };
  });
  return /*#__PURE__*/React.createElement("div", {
    ref: container
  }, children);
});
MagicGridWrapper.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node)
};

export default MagicGridWrapper;
